1-Draft/RFC00XX-Module-Isolation.md

```diff
@@ -0,0 +1,251 @@
+
+### Caching Problem
+
+PowerShell has a lot caches related to `System.Type`.
+Some have `Type` as the value, such as the `TypeCache` in `TypeResolver` and the cached type for `ITypeName`.
+Some have `Type` as the key, such as the member tables in `DotNetAdapter` and the binder caches in DLR.
+
+For the former category, as was already mentioned above,
+the cache data structure needs to be updated so as to make sure the cache is associated with a load context.
+
+  - For example, when type resolution happens for `[C]` that is running from an isolated module M,
+    then `TypeResolver` should use the cache associated with ALC-M.
+
+  - Another example, if a `ITypeName` has already resolved the type via reflection in default-ALC,
+    and the `ScriptBlock` containing that `ITypeName` is now being executed in the context of an isolated module M,
+    then the cached type referred by `ITypeName` should be invalided and another resolution should be triggered.
+    Similarly, if the `ScriptBlock` has already been compiled into LINQ expression tree and have the delegate generated,
+    then the cached delegate needs to be voided and another compilation needs to be enforced to make sure type resolution is correct.
+
+For the latter category, theoretically they will continue to work without change,
+because types loaded into different load contexts are different.
+
+### Reclaiming Module AssemblyLoadContext
+
+Starting from 3.0, .NET Core supports unloading an `AssemblyLoadContext` and all assemblies in it.
+Ideally, we would like to reclaim the load context associated with an isolated module when the module is removed.
+
+The cache updates called out in the [Update Type Resolution Logic](#update-type-resolution-logic) section
+were designed with the purpose of allowing the custom ALC to be reclaimed.
+By having each custom ALC to hold the cache data structure within itself,
+we won't need to hold a custom ALC or any types loaded in a custom ALC as the key or value for any caches in our engine code,
+and thus there is no need to invalidate any engine cache.
+
+However, we won't be able to guarantee that a custom ALC could be unloaded at all.
+This is because Dynamic Language Runtime (DLR) holds types in its cache, indefinitely.
+
+The DLR types live in the default-ALC,
+and it will generate a `DynamicMetaObject` object for a late-binding operation,
+which is something like a key-value pair `<restrictions, expression>` that will be held in the DLR caches.
+This is a perf improvement feature for dynamic languages like PowerShell because it allows you to skip an expensive late-binding operation when the `restrictions` matches.
+That key-value pair will be held indefinitely,
+so once types from a custom load context is involved in a late binding operation,
+such as property access, method invocation and etc,
+it will be rooted in the GC Heap by the DLR cache, and the custom load context cannot be unloaded afterwards.
+
+## Summary
+
```

Comment:
I am wondering about the debugger experience in this new world. Do you have any thoughts on this?

Reply:
I don't think there will be any changes in the debugging experience. Behavior-wise, only the type resolution and possible type identity issues are user visible, but they shouldn't affect the debugging experience.